lagrangebench/evaluate/rollout.py

```diff
@@ -54,64 +102,80 @@ def eval_single_rollout(
+        # 1. preprocess features
+        features_batch, neighbors_batch = preprocess_eval_vmap(
+            sample_batch, neighbors_batch
+        )
 
-            continue
+        # 2. check whether list overflowed and fix it if so
+        if neighbors_batch.did_buffer_overflow.sum() > 0:
+            # check if the neighbor list is too small for any of the samples
+            # if so, reallocate the neighbor list
 
-        # predict
-        pred, _ = model_apply(params, state, (features, particle_type))
+            print(f"(eval) Reallocate neighbors list at step {step}")
+            ind = jnp.argmax(neighbors_batch.did_buffer_overflow)
+            sample = broadcast_from_batch(sample_batch, index=ind)
 
-        next_position = case.integrate(pred, current_positions)
+            _, nbrs_temp = case.allocate_eval(sample)
+            print(
+                f"(eval) From {neighbors_batch.idx[ind].shape} to {nbrs_temp.idx.shape}"
+            )
+            neighbors_batch = broadcast_to_batch(nbrs_temp, batch_size)
 
-        if n_extrap_steps == 0:
-            kinematic_mask = get_kinematic_mask(particle_type)
-            next_position_ground_truth = ground_truth_positions[:, step]
+            # To run the loop N times even if sometimes
+            # did_buffer_overflow > 0 we directly return to the beginning
 
-            next_position = jnp.where(
-                kinematic_mask[:, None],
-                next_position_ground_truth,
-                next_position,
-            )
-        else:
-            warnings.warn("kinematic mask not applied in extrapolation mode.")
+            continue
 
-        predictions = predictions.at[step].set(next_position)
-        current_positions = jnp.concatenate(
-            [current_positions[:, 1:], next_position[:, None, :]], axis=1
+        # 3. run forward model
+        current_positions_batch = forward_eval_vmap(
+            params,
+            state,
+            (features_batch, particle_type_batch),
+            current_positions_batch,
+            target_positions_batch[:, :, step],
+        )
+
+        # 4. write predicted next position to output array
+        predictions_batch = predictions_batch.at[:, step].set(
+            current_positions_batch[:, :, -1]  # most recently predicted positions
         )
 
         step += 1
 
-    # (n_nodes, traj_len - t_window, dim) -> (traj_len - t_window, n_nodes, dim)
-    ground_truth_positions = ground_truth_positions.transpose(1, 0, 2)
+    # (batch, n_nodes, time, dim) -> (batch, time, n_nodes, dim)
+    target_positions_batch = target_positions_batch.transpose(0, 2, 1, 3)
+    metrics_batch = vmap(metrics_computer)(predictions_batch, target_positions_batch)
 
-    return (
-        predictions,
-        metrics_computer(predictions, ground_truth_positions),
-        neighbors,
-    )
+    return (predictions_batch, metrics_batch, broadcast_from_batch(neighbors_batch, 0))
```

Comment:
I might be missing something on what is contained in `neighbors_batch`. Why do you return only `neighbors_batch[0]`?

Reply:
One could move batching the neighbors to `eval_rollout` instead of doing it every time within eval_batched_rollout. Shouldn't change anything to the functionality.

Reply:
I realized that batch_size might be a value smaller than the batch size in the configs if the last batch is not full. And within `eval_batched_rollout` it is the current batch_size, not the config one. I now moved out everything I could from `eval_batched_rollout` and left only the parts that depend on the current_batch_size. I also added a comment clarifying that part.
Thanks for the hint!